Fluent UI v2 CSS provider

stsrki opened the pull request:
Closes #4135
Closes #5287
Closes #5292

stsrki commented:
@David-Moreira please review once you get the chance.

David-Moreira commented:
> @David-Moreira please review once you get the chance.

![image](https://github.com/Megabit/Blazorise/assets/22283161/c0e6c8e8-6a0d-45ea-a14a-22316cff29aa)

LGTM...?? hahaha

stsrki commented:
> > @David-Moreira please review once you get the chance.
> 
> ![image](https://private-user-images.githubusercontent.com/22283161/303871357-c0e6c8e8-6a0d-45ea-a14a-22316cff29aa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc2NDI4NzEsIm5iZiI6MTcwNzY0MjU3MSwicGF0aCI6Ii8yMjI4MzE2MS8zMDM4NzEzNTctYzBlNmM4ZTgtNmEwZC00NWVhLWExNGEtMjIzMTZjZmYyOWFhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjExVDA5MDkzMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE1ZTMyNGIwOGE0ZDA2MWEyMjhhODMyMThmNmFhM2Y5OWY3NTRmNGQzOTgxMjExZWFkNTBlNGMxYjdiZDY4NjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.rvs2miIAa0wR6TVlL07_rCReKR3i3e9CJwmixhBJsMM)
> 
> LGTM...?? hahaha

Pfff, that's nothing 🤣

David-Moreira commented:
**UI Review based on our Demo:**

**Disclaimer :** I have never ever used Fluent UI neither am well acquainted with it, so I can't comment whether something looks like it should in Fluent UI. I will comment things that I think might be wrong or done better.

1. Header should stay on top, the body also seems to be cut off, it's behind the header
![image](https://github.com/Megabit/Blazorise/assets/22283161/77971d7e-3ee1-4b7f-8413-6004436688c6)

2. Theme settings seem to not have any effect?
![image](https://github.com/Megabit/Blazorise/assets/22283161/910457cc-3a48-465b-af4a-95426541fdcd)

3. The mouse hover here, does not display a clickable indicator, is this expected?
![image](https://github.com/Megabit/Blazorise/assets/22283161/82097fba-9498-4583-bb5f-da3ec96d9bf4)

4. Is it expected that the alerts are hidden instead of removed from the html? For example, bs seems to regain back the space once the alerts are closed.
![image](https://github.com/Megabit/Blazorise/assets/22283161/c68e0d42-be0e-4fcb-b1f3-7255ba39ce34)
![image](https://github.com/Megabit/Blazorise/assets/22283161/95e12577-22b7-470f-8271-f3dda829fd2b)

5. Do the menu items not have a selected / active look?
![image](https://github.com/Megabit/Blazorise/assets/22283161/3bb889c3-e814-4d56-a520-08a2d6876afc)

6. Does fluent UI use explicit text for previous/next? Does this needs to be localized?
![image](https://github.com/Megabit/Blazorise/assets/22283161/7195914e-3fe4-4885-ad3d-676c4a10fa88)

7. The Modal Animation for Fade In does not seem to obbey animation duration

8. The pagination has no hover look or anything, seems very static, is this expected? Also it should be visible which pages are selected/active.

9. The Danger Page Progress does not seem to be different.

10. Table Fixed Columns seems to be broken
![image](https://github.com/Megabit/Blazorise/assets/22283161/e73a8774-075b-450f-a7ab-588ae7d8caad)

11. Should Addon be this big?
![image](https://github.com/Megabit/Blazorise/assets/22283161/cd5a5cb5-e285-40d3-8425-0a54ad5e96f0)

12. Table Resize feature seems to be broken, both Header And Columns

13. The Faded tooltip looks like the others?

14. The naming should actually be Start/End instead of Left/Right?
![image](https://github.com/Megabit/Blazorise/assets/22283161/d39fdc16-b12f-4c93-9a72-a517823f3791)

15. Should the 100 represent the currently selected value, or the maximum value?
![image](https://github.com/Megabit/Blazorise/assets/22283161/d5847869-bddf-4054-b152-26848b51f664)

16. No radio buttons
![image](https://github.com/Megabit/Blazorise/assets/22283161/184784d4-a9e6-4900-8bb3-35369890f85a)

17. No color at all
![image](https://github.com/Megabit/Blazorise/assets/22283161/dc51d6d5-264b-40fc-9dbc-5a8b54141074)

18. Is this how addons look? I can't really tell the Start apart from the text. Also not enough spacing? Is this how fluent UI states it should be like?
![image](https://github.com/Megabit/Blazorise/assets/22283161/1d9f16d7-485a-4870-adf9-66c66277c6d4)

19. I guess a Dropdown might not be natively supported as addon, but maybe it should not have the left padding?
![image](https://github.com/Megabit/Blazorise/assets/22283161/f24943c9-3e24-4f92-8f9a-4d45247ab536)

20. Addon button right padding?
![image](https://github.com/Megabit/Blazorise/assets/22283161/1d6b4abd-4ec3-4b90-9efe-6ff128a3ea0f)

21. Some addons don't look very good
![image](https://github.com/Megabit/Blazorise/assets/22283161/6e72b6b5-fb75-4675-b3b6-06229bb6ccec)

22. Why does the DataGrid icon look like a music note?
![image](https://github.com/Megabit/Blazorise/assets/22283161/833517e1-8586-423b-81e8-f0982b752d68)

23. Pagination in the DataGrid : 
- I need to know which page is selected
- Previous/Next First/Last buttons should appear as disabled accordingly
- The dropdown page selection should only appear in certain resolutions?
![image](https://github.com/Megabit/Blazorise/assets/22283161/cff5c8da-412a-45c2-aecf-00f1638955d3)

24. Datagrid Column filters don't look too good
![image](https://github.com/Megabit/Blazorise/assets/22283161/b7b7dd7f-30e5-4811-bc5e-cbb4afa9e8ba)

25. DataGrid Fixed Header too big?
![image](https://github.com/Megabit/Blazorise/assets/22283161/fc4cfc4d-2c87-43f4-b723-c0ab94086472)

26. Pagination look. Some icons don't seem correct?
![image](https://github.com/Megabit/Blazorise/assets/22283161/d7675764-d022-4acc-9f2f-91888898cf20)

27. DataGrid Resize mode, also broken (table related) http://localhost:5138/tests/datagrid/resize

28. Fixed Columns create a white background on top of selected look
![image](https://github.com/Megabit/Blazorise/assets/22283161/1e8432f4-726c-4e45-8f06-41fb9410295a)

29. DataGrid Multiple : Clicking the checkbox does not check it
![image](https://github.com/Megabit/Blazorise/assets/22283161/fb6cbb92-dbc3-4067-a025-53c02555c9cd)

30. Datagrid Popup Edit with horizontal bar (overflow)
![image](https://github.com/Megabit/Blazorise/assets/22283161/3b1d0139-aa6b-4350-88f9-83a2af55bc69)

31: DataGrid Header looks too big (It's from resize feature being activated apparently)
![image](https://github.com/Megabit/Blazorise/assets/22283161/c791d7ed-4034-44af-88ff-9911a86e78d3)

32. Autocomplete selection does not look very good
![image](https://github.com/Megabit/Blazorise/assets/22283161/387d8680-944b-4c2a-98db-3036e7a1f600)

33. I believe the dropdown list has max height or something. There should be a scroll
![image](https://github.com/Megabit/Blazorise/assets/22283161/87cb81fd-d042-4553-8a69-19b6a85e0f16)

34. The top snackbars appear behind the header
http://localhost:5138/tests/snackbar

35. Are all the borders correct?
![image](https://github.com/Megabit/Blazorise/assets/22283161/bc3c42e7-8244-43f8-a804-cd74511afc5e)

stsrki commented:
Work so far

1. fixes
2. I have planned for this. We will skip it for noe.
3. fixed
4. fixed
5. fixed
6. working as expected, and we also have localization
7. 
8. fixed
9. fixed
10. David
11. demo ui adjusted
12. David
13. fixed
14. fixed
15. yes, the value shows min and max acording to fluent design
16. 
17. 
18. buttons inside of addons asjusted to not have colors, acording to fluent design
19. save as 18. Padding must be there acording to fluent design
20. same as 18. and 19.
21. as I don't have a reference for diferent addon sizes I had to improvise. Until we know more this is how it works.
22. fixed. introduces new Table icon
23. fixed
24. 
25. fixed by adding missing CSS
26. these icons are the best I could find in the Fluent UI icons package. If you want, you can look at https://react.fluentui.dev/?path=/docs/concepts-developer-icons-icons-catalog--page for alternative
27. fixed with 25.
28. 
29. 
30. 
31. fixed by adding missing CSS
32. David
33. fixed
34. 
35.

David-Moreira commented on Source/Blazorise.FluentUI2/Styles/components/_card.scss:
```diff
@@ -5,8 +5,6 @@
     display: flex;
     flex-direction: column;
-    overflow-x: hidden;
-    overflow-y: hidden;
```
@stsrki Please evaluate this change. 
The problem is if the overflow is not permitted, inner content will not be available to overflow on top of the card.

David-Moreira replied:
(Autocomple dropdown fix)

stsrki replied:
What is the problem here? Can you clarify?

David-Moreira replied:
Did I not just say? 
![image](https://github.com/Megabit/Blazorise/assets/22283161/bc9bbda3-17f7-4b8d-8784-c7435f171a86)

stsrki replied:
Yes, but I tried it on my side, and it works. That's why I'm asking for more information.

David-Moreira replied:
It works due to this change

stsrki replied:
It seems it is not working after all.

![image](https://github.com/Megabit/Blazorise/assets/900302/9cfe0883-a2c8-4bf0-a14d-1202004bac2a)

I will look into it.

David-Moreira replied:
Ok, so I fixed Autocomplete, and I broke that one then?
Please test Autocomplete with whatever changes you make.

stsrki replied:
It should be fixed now. Autocomplete also works.

stsrki commented:
On Autocomplete, we still show checked items with an icon.

![image](https://github.com/Megabit/Blazorise/assets/900302/62826dff-8493-4ce5-b8c6-ea8a84724050)

Now that we have real checkboxes on DropdownItem it would look much better with real checkboxes.

![image](https://github.com/Megabit/Blazorise/assets/900302/eedbc420-1fbf-4f64-9243-3fceb1558b75)

![image](https://github.com/Megabit/Blazorise/assets/900302/1b8c375a-4d27-4e1f-9c6c-ee4da3dfcab8)

---

@David-Moreira Would you like to work on it? In this PR or should we open new ticket once we merge this Fluent UI?

stsrki commented:
Leaving what was left to do

**UI Review based on our Demo:**

10. Table Fixed Columns seems to be broken
![image](https://github.com/Megabit/Blazorise/assets/22283161/e73a8774-075b-450f-a7ab-588ae7d8caad)

24. Datagrid Column filters don't look too good
![image](https://github.com/Megabit/Blazorise/assets/22283161/b7b7dd7f-30e5-4811-bc5e-cbb4afa9e8ba)

29. DataGrid Multiple : Clicking the checkbox does not check it
![image](https://github.com/Megabit/Blazorise/assets/22283161/fb6cbb92-dbc3-4067-a025-53c02555c9cd)

David-Moreira commented:
**12.** Fixed 

Can you do **11**? You probably better suited since it's styling. And you're already used to the fluent styling?

David-Moreira commented on Source/Extensions/Blazorise.DataGrid/_DataGridPagination.razor:
```diff
@@ -175,7 +175,7 @@
                     var pageActive = pageNumber == PaginationContext.CurrentPage;
 
-                    <PaginationItem Display="Blazorise.Display.None.OnMobile.InlineBlock.OnTablet" Disabled="@(pageNumber == PaginationContext.CurrentPage)" Active="@pageActive">
+                    <PaginationItem Display="Blazorise.Display.None.OnMobile.InlineFlex.Row.OnTablet" Disabled="@(pageNumber == PaginationContext.CurrentPage)" Active="@pageActive">
```
This is a wide change, is this ok for all providers?

stsrki replied:
Yes, I tried in all providers. Works without visible issues.

stsrki commented:
> **12.** Fixed
> 
> Can you do **11**? You probably better suited since it's styling. And you're already used to the fluent styling?

Much better, I think

![image](https://github.com/Megabit/Blazorise/assets/900302/202a3eab-e573-4373-9dc8-c47114e7640b)

David-Moreira commented:
> > **12.** Fixed
> > Can you do **11**? You probably better suited since it's styling. And you're already used to the fluent styling?
> 
> Much better, I think
> 
> ![image](https://private-user-images.githubusercontent.com/900302/305172216-202a3eab-e573-4373-9dc8-c47114e7640b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDgwMjAzNjUsIm5iZiI6MTcwODAyMDA2NSwicGF0aCI6Ii85MDAzMDIvMzA1MTcyMjE2LTIwMmEzZWFiLWU1NzMtNDM3My05ZGM4LWM0NzExNGU3NjQwYi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjE1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIxNVQxODAxMDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lM2IwNDkyNTgyNDE1YjhlNjMzMzNiNmE1ZDNhNDYzNDNjY2QyMTU5NzkzZGE2YzMzMTg2NzFlYjM5YTJkNzJjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.B4dYXZEFGMo9962KEG9mPpNf0qeQDzDPycZhI5SnRrY)

Yep, you the man!!

stsrki commented:
I'm eager to merge. Wanna QA fast and not tell me a thing? 🤣🤣

David-Moreira commented:
1. I still see no visible changes when trying to change the theme?
![image](https://github.com/Megabit/Blazorise/assets/22283161/1f047e00-8faa-4d78-9846-3bb2acf04ec4)

2. Is this the expected Modal look? Doesn't look like the header & footer are noticeably separate from the body.

![image](https://github.com/Megabit/Blazorise/assets/22283161/fd9e4c85-7566-4bac-a4f2-59c6a88106e8)

3. The pagination has no hover look, usually clickable / interactible items have some visual hint to them. Does fluent ui not have that? Regular buttons seem to have it.

4. I don't see anything in the 2 pixels?

![image](https://github.com/Megabit/Blazorise/assets/22283161/a14483fa-fe0d-4c09-a37b-0bc449a8e0e1)

All in all, everything acts and looks better to me.

David-Moreira commented on Demos/Blazorise.Demo.FluentUI2/Blazorise.Demo.FluentUI2.csproj:
```diff
@@ -0,0 +1,22 @@
+    <ProjectReference Include="..\Blazorise.Demo\Blazorise.Demo.csproj" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <Content Update="_Imports.razor">
+      <ExcludeFromSingleFile>true</ExcludeFromSingleFile>
+      <Generator>MSBuild:RazorGenerateComponentDeclarationDesignTime</Generator>
+    </Content>
+  </ItemGroup>
+
+</Project>
```
Should this be here?

David-Moreira commented on Demos/Blazorise.Demo.FluentUI2/dotnetwatchrun.bat:
```diff
@@ -0,0 +1,6 @@
+: This is here to enable console colors when using watch
+: Usefull for debugging when developping
+set DOTNET_WATCH_SUPPRESS_LAUNCH_BROWSER=1
+title Blazorise.Demo.Bulma
```
wrong title

David-Moreira commented on Source/Blazorise.FluentUI2/Components/AlertMessage.razor:
```diff
@@ -0,0 +1,27 @@
+﻿@inherits Blazorise.AlertMessage
+<div @ref="@ElementRef" id="@ElementId" class="@ClassNames" style="@StyleNames" @attributes="@Attributes">
+    @* @if ( ParentAlert?.Color is not null && ParentAlert?.Color != Color.Default )
+    {
+        <div class="fui-MessageBar__icon">
+            @if ( ParentAlert?.Color == Color.Success )
+            {
+                <svg fill="currentColor" aria-hidden="true" width="1em" height="1em" viewBox="0 0 20 20" xmlns="http://www.w3.org/2000/svg"><path d="M10 2a8 8 0 1 1 0 16 8 8 0 0 1 0-16Zm3.36 5.65a.5.5 0 0 0-.64-.06l-.07.06L9 11.3 7.35 9.65l-.07-.06a.5.5 0 0 0-.7.7l.07.07 2 2 .07.06c.17.11.4.11.56 0l.07-.06 4-4 .07-.08a.5.5 0 0 0-.06-.63Z" fill="currentColor"></path></svg>
+            }
+            else if ( ParentAlert?.Color == Color.Danger )
+            {
+                <svg fill="currentColor" aria-hidden="true" width="1em" height="1em" viewBox="0 0 20 20" xmlns="http://www.w3.org/2000/svg"><path d="M10 2a8 8 0 1 1 0 16 8 8 0 0 1 0-16Zm0 10.5a.75.75 0 1 0 0 1.5.75.75 0 0 0 0-1.5ZM10 6a.5.5 0 0 0-.5.41v4.68a.5.5 0 0 0 1 0V6.41A.5.5 0 0 0 10 6Z" fill="currentColor"></path></svg>
+            }
+            else if ( ParentAlert?.Color == Color.Warning )
+            {
+                <svg fill="currentColor" aria-hidden="true" width="1em" height="1em" viewBox="0 0 20 20" xmlns="http://www.w3.org/2000/svg"><path d="M8.68 2.79a1.5 1.5 0 0 1 2.64 0l6.5 12A1.5 1.5 0 0 1 16.5 17h-13a1.5 1.5 0 0 1-1.32-2.21l6.5-12ZM10.5 7.5a.5.5 0 0 0-1 0v4a.5.5 0 0 0 1 0v-4Zm.25 6.25a.75.75 0 1 0-1.5 0 .75.75 0 0 0 1.5 0Z" fill="currentColor"></path></svg>
+            }
+            else 
+            {
+                <svg fill="currentColor" aria-hidden="true" width="1em" height="1em" viewBox="0 0 20 20" xmlns="http://www.w3.org/2000/svg"><path d="M18 10a8 8 0 1 0-16 0 8 8 0 0 0 16 0ZM9.5 8.91a.5.5 0 0 1 1 0V13.6a.5.5 0 0 1-1 0V8.9Zm-.25-2.16a.75.75 0 1 1 1.5 0 .75.75 0 0 1-1.5 0Z" fill="currentColor"></path></svg>
+            }
+        </div>
+    } *@
```
Is this all commented?

David-Moreira reviewed (approved):
LGTM, most of the files I glanced over. Too many files...

stsrki commented:
1. as I have mentioned before. For now I'm not gonna do the theme generator. As an alternative, users could use Fluent design tokens, which I will document before release.
2. Yes, this is how it looks natively. In fact it works even worse. In our version we at least have some spacings.
3. Fixed it
4. This works, but since Secondary color has a low contrast compared to background it is not so much visible.

stsrki commented:
Regarding the theme generator. Basically, this is my idea:

- Goto https://react.fluentui.dev/?path=/docs/theme-theme-designer--page
- Configure theme
- Export
- Copy tokens
- Apply tokens in Blazorise somehow
- Generate CSS variables from JSON tokens

![image](https://github.com/Megabit/Blazorise/assets/900302/0c912c68-c4a0-4808-a77d-a03ac453d609)

---

That is how it is done manually.

We can expand from there and automate it a little. Generate a C# model from the JSON object with all of the token fields. Then, in Blazorise, the theme generator applies colors and other variables to regenerate new JSON object, and in turn, it generates CSS new variables.

David-Moreira commented:
I see ok than. I guess we can improve that later.
